internal/hcsoci/credentials.go

```diff
@@ -0,0 +1,137 @@
+	// gets passed in. Transport either "LRPC" (Argon) or "HvSocket" (Xenon).
+	// 2. Query the instance with a call to HcsGetServiceProperties with the
+	// PropertyType "ContainerCredentialGuard". This will return all instances
+	// 3. Parse for the id of our container to find which one correlates to the
+	// container we're building the doc for, then add to the v2 doc.
+	// 4. If xenon container the hvsocketconfig will need to be attached BEFORE
+	// being created. It must be in the doc itself as we do not support hot adding
+	// service table entries. This is currently a blocker for adding support for
+	// hyper-v gmsa.
+	transport := "LRPC"
+	if uvm {
+		transport = "HvSocket"
+	}
+	req := hcsschema.ModificationRequest{
+		PropertyType: "ContainerCredentialGuard",
+		Settings: &hcsschema.ContainerCredentialGuardOperationRequest{
+			Operation: hcsschema.AddInstance,
+			OperationDetails: &hcsschema.ContainerCredentialGuardAddInstanceRequest{
+				Id:             ID,
+				CredentialSpec: credSpec,
+				Transport:      transport,
+			},
+		},
+	}
+	if err := hcs.ModifyServiceSettings(ctx, req); err != nil {
+		log.G(ctx).WithError(err).WithField("containerID", ID).Warn("failed to generate container credential guard instance")
+		return nil, err
+	}
+
+	q := hcsschema.PropertyQuery{
+		PropertyTypes: []hcsschema.PropertyType{hcsschema.PTContainerCredentialGuard},
+	}
+	serviceProps, err := hcs.GetServiceProperties(ctx, q)
+	if err != nil {
+		log.G(ctx).WithError(err).WithField("containerID", ID).Warn("failed to retrieve container credential guard instances")
+		return nil, err
+	}
+
+	ccgState, err := unmarshalCCGInstances(serviceProps, ID)
+	if err != nil {
+		return nil, err
+	}
+	return ccgState, nil
+}
+
+// Checks to see if doc has non-nil ccg field
+func hasCCG(doc *hcsschema.Container) bool {
+	if doc.ContainerCredentialGuard != nil {
+		return true
+	}
+	return false
+}
+
+// Removes a ContainerCredentialGuard instance by container ID.
+func removeCredentialGuard(ctx context.Context, ID string) error {
+	log.G(ctx).WithField("containerID", ID).Debug("removing container credential guard")
+
+	req := hcsschema.ModificationRequest{
+		PropertyType: "ContainerCredentialGuard",
+		Settings: &hcsschema.ContainerCredentialGuardOperationRequest{
+			Operation: hcsschema.RemoveInstance,
+			OperationDetails: &hcsschema.ContainerCredentialGuardRemoveInstanceRequest{
+				Id: ID,
+			},
+		},
+	}
+	if err := hcs.ModifyServiceSettings(ctx, req); err != nil {
+		return err
+	}
+	return nil
+}
+
+// Unmarshals a ServiceProperties struct if the only property queried was 'ContainerCredentialGuard'
+// Returns the ContainerCredentialGuardInstance matching the id passed in.
+func unmarshalCCGInstances(sp *hcsschema.ServiceProperties, ID string) (*hcsschema.ContainerCredentialGuardState, error) {
+	if len(sp.Properties) != 1 {
+		return nil, fmt.Errorf("wrong number of service properties present")
+	}
+	// Properties is []interface{}
```

Comment:
Can we make Properties be []RawJson instead?

Reply:
Do you mean RawMessage? And yep that seems better suited, I didn't know if we usually edit the generated schema files to more appropriate types so I left what it outputted mostly.